### ISSUE TITLE
IDO donwtimehistory records orphaned from scheduleddowntime records following restart

### ISSUE THREAD

**Issue by mdetrano** [Icinga/icinga2#5458]
Following a restart of Icinga2, records in the the downtimehistory table may have internal_downtime_id's that no longer match the correct record in the scheduleddowntime table.  As a result, these records do not get updated correctly when the Downtime ends or is cancelled.

## Expected Behavior

The downtimehistory records should get updated correctly (actual_end_time set, is_in_effect updated, was_cancelled, etc) regardless of whether a restart/reload of Icinga2 was performed prior to the Downtime period.

## Current Behavior

The history table may indicate that downtime started, but never ended, despite the fact that the object is no longer in downtime.

## Possible Solution

I've implemented a very rough workaround by running a cron job and checking that the object_id fields in scheduleddowntime, and downtimehistory records match up... and updating the internal_downtime_id field in the history table accordingly when they don't.  It seems to me that the internal_downtime_id values should never get re-used, and simply be carried through a restart of the system.

## Steps to Reproduce (for bugs)

1. Schedule downtime for several objects in the system (more than one).
2. Restart icinga2
3. Cancel the downtimes or let them expire.
4. Check downtime history in IDO database or via a web interface (like Icinga Web 2)... some objects will show downtime starting, but never ending.

## Context

Not having accurate downtime history recorded in the database makes it hard to do availability reporting where downtime states a segregated out for informational purposes.  (problems during planned downtime often aren't supposed to affect overall SLA calculations, for example).

## Your Environment

* Version used (`icinga2 --version`): 2.7.0
* Operating System and version: Centos &
* Enabled features (`icinga2 feature list`):
api checker command ido-mysql livestatus mainlog
* Icinga Web 2 version and modules (System - About):
* Config validation (`icinga2 daemon -C`):
* If you run multiple Icinga 2 instances, the `zones.conf` file (or `icinga2 object list --type Endpoint` and `icinga2 object list --type Zone`) from all affected nodes.

**Comment by dnsmichi** [comment]
Could be a similar issue as #5456 @Crunsher @gunnarbeutner

**Comment by dnsmichi** [comment]
Hmmm, I would guess that the legacy_id is empty or wrong after the downtime is deleted/cancelled which renders the WHERE condition useless.

Still guessing why that would happen, one thing could be a purged icinga2.state file which deletes the stored legacy_id. 

Breaking up the condition is dangerous, as there is no index. It will partially match the unique key defined on the downtimehistory table. 

```
diff --git a/lib/db_ido/dbevents.cpp b/lib/db_ido/dbevents.cpp
index ac505aaa6..a92bee2a3 100644
--- a/lib/db_ido/dbevents.cpp
+++ b/lib/db_ido/dbevents.cpp
@@ -612,7 +612,6 @@ void DbEvents::RemoveDowntimeInternal(std::vector<DbQuery>& queries, const Downt

        query3.WhereCriteria = new Dictionary();
        query3.WhereCriteria->Set("object_id", checkable);
-       query3.WhereCriteria->Set("internal_downtime_id", downtime->GetLegacyId());
        query3.WhereCriteria->Set("entry_time", DbValue::FromTimestamp(downtime->GetEntryTime()));
        query3.WhereCriteria->Set("instance_id", 0); /* DbConnection class fills in real ID */
```

Furthermore the unique relationship will be object_id and entry_time. If you have two downtimes in the same second, the update will be wrong.

We could instead use the `name` column, but TEXT compares are really expensive.

@mdetrano try the patch in a test lab please.

@Crunsher @N-o-X this seems to precisely describe the issue we've seen in the meeting last Wednesday.

@lippserd any further ideas?

**Comment by Crunsher** [comment]
That very much sounds like issue we were facing, we can reproduce this... but which downtimes are handled and deleted correctly and which are not seems to be completely random.

I saw results ranging from 2 out of 10 to 50 out of 50 downtimes not being deleted correctly

**Comment by mdetrano** [comment]
The patch seems to work for at least getting the end times updated on the downtime_history records.  Some of the other fields didn't get updated (was_started, is_in_effect).  I tested both cancelling the downtimes, and letting them expire (after performing restarts following downtime creation and verifying that the internal_ids no longer matched up).  It looks like the general idea is correct---just skip the internal_downtime_id when hunting for matching records for the downtime.

**Comment by dnsmichi** [comment]
`was_started` is only updated if the downtime starts/is triggered. That's a different event and happens independent from deleting/cancelling the downtime. `is_in_effect` will be set to 0. `actual_end_time` is only updated for

- a fixed downtime
- a flexible downtime which has been triggered. If it wasn't triggered, it did not even start yet.

TL;DR - removing `internal_downtime_id` is not the 100% patch, we still need a unique `name` for the where condition. Otherwise we would update 2..n downtimes for one checkable which where added in the same second (entry_time).

**Comment by Crunsher** [comment]
Problem:
* legacy_downtime is not a persistent value
* removing it does not solve the issue as object_id + entry_time are not unique

Possible Solution:
Filter by name, which is always unique... but can't be used as an index since it's a TEXT field. Adding an index on object_id, entry_time, instance_id and using name as an additional WHERE does the trick, but would require a db scheme update.

```
MariaDB [icinga]> explain select * from icinga_downtimehistory WHERE object_id = 211 AND entry_time = "2017-08-04 16:59:12" AND "instance_id" = 0 AND name = "jfws!jfws-1501858752-9";
+------+-------------+------------------------+------+----------------+------+---------+------+------+-------------+
| id   | select_type | table                  | type | possible_keys  | key  | key_len | ref  | rows | Extra       |
+------+-------------+------------------------+------+----------------+------+---------+------+------+-------------+
|    1 | SIMPLE      | icinga_downtimehistory | ALL  | sla_idx_dohist | NULL | NULL    | NULL | 1163 | Using where |
+------+-------------+------------------------+------+----------------+------+---------+------+------+-------------+
1 row in set, 1 warning (0.00 sec)

MariaDB [icinga]> CREATE index bombel on icinga_downtimehistory (object_id, entry_time, instance_id);
Query OK, 0 rows affected (0.04 sec)
Records: 0  Duplicates: 0  Warnings: 0

MariaDB [icinga]> explain select * from icinga_downtimehistory WHERE object_id = 211 AND entry_time = "2017-08-04 16:59:12" AND "instance_id" = 0 AND name = "jfws!jfws-1501858752-9";
+------+-------------+------------------------+------+-----------------------+--------+---------+-------------+------+-------------+
| id   | select_type | table                  | type | possible_keys         | key    | key_len | ref         | rows | Extra       |
+------+-------------+------------------------+------+-----------------------+--------+---------+-------------+------+-------------+
|    1 | SIMPLE      | icinga_downtimehistory | ref  | sla_idx_dohist,bombel | bombel | 14      | const,const |    4 | Using where |
+------+-------------+------------------------+------+-----------------------+--------+---------+-------------+------+-------------+
1 row in set, 1 warning (0.00 sec)

**Comment by Crunsher** [comment]
2.7. fix: Remove internal downtime id from query and use scheduled_start_time, scheduled_end_time, is_in_effect. And create an index for those

2.8 fix: Change name to VARCHAR and create an index on that. This change would also make sense for the comment_history

**Comment by Crunsher** [comment]
The PR fixes this problem (tested for mysql, pending for pgsql).

But this mismatch can also happen for comments.

**Comment by dnsmichi** [comment]
I'll leave that as working ticket for 2.8.0.

**Comment by xenuser** [comment]
@mdetrano Would you mind sharing your workaround somewhere? Would highly appreciate it!

**Comment by mdetrano** [comment]
@xenuser , I think that latest 2.7.1 version of Icinga2 now has this issue fixed...I suspect doing an update will be simpler than the kludge I made.

**Comment by Crunsher** [comment]
The original issue is fixed in 2.7.1 thanks to a change of our query. We want to extend our hackish solution in 2.8.0 by altering the db-scheme but no need to worry about that now~
